File: 0003 Longest Substring Without Repeating Characters/LongestSubstringWithoutRepeatingCharacters.cpp
#define MAX(a, b) (a > b) ? a : b
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int s_len = s.length();
        if (s_len <= 1)
            return s_len;
        
        int head = 0, index = 0, ret = 0;
        int map[256] = {0};
        
        while (index < s_len) {
            char c = s.at(index);
            if (map[c] != 0) {
                ret = MAX(index - head, ret);
                head = MAX(head, map[c]);

                if (s_len - head < ret)
                    return ret;
            }
            map[c] = ++index;
        }
        return MAX(s_len - head, ret);
    }
};

File: 0003 Longest Substring Without Repeating Characters/LongestSubstringWithoutRepeatingCharacters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        s_len = len(s)
        if s_len <= 1:
            return s_len

        head = 0
        ret = 0
        dic = dict()

        for index, c in enumerate(s):
            if c in dic:
                ret = max(index - head, ret)
                head = max(dic[c] ,head)

                if s_len - head < ret:
                    return ret
            dic[c] = index + 1
        return max(s_len - head ,ret)
